server/src/main/java/org/opensearch/index/ShardIndexingPressureStore.java

```diff
@@ -0,0 +1,121 @@
+ * object is brought into the hot store again, until it remains active. If not present in the either store, a fresh
+ * object is instantiated and registered in both the stores for concurrent accesses.
+ *
+ * Note: The implementation of shardIndexingPressureColdStore methods is such that get,
+ * update and evict operations can be abstracted out to support any other strategy such as LRU, if
+ * discovered a need later.
+ *
+ */
+public class ShardIndexingPressureStore {
+
+    // This represents the maximum value for the cold store size.
+    public static final Setting<Integer> MAX_COLD_STORE_SIZE =
+        Setting.intSetting("shard_indexing_pressure.cache_store.max_size", 200, 100, 1000,
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
+
+    private final Map<ShardId, ShardIndexingPressureTracker> shardIndexingPressureHotStore =
+        ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
+    private final Map<ShardId, ShardIndexingPressureTracker> shardIndexingPressureColdStore =
+        ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
+    private final ShardIndexingPressureSettings shardIndexingPressureSettings;
+
+    private volatile int maxColdStoreSize;
+
+    public ShardIndexingPressureStore(ShardIndexingPressureSettings shardIndexingPressureSettings,
+                                      ClusterSettings clusterSettings, Settings settings) {
+        this.shardIndexingPressureSettings = shardIndexingPressureSettings;
+        this.maxColdStoreSize = MAX_COLD_STORE_SIZE.get(settings).intValue();
+        clusterSettings.addSettingsUpdateConsumer(MAX_COLD_STORE_SIZE, this::setMaxColdStoreSize);
+    }
+
+    public ShardIndexingPressureTracker getShardIndexingPressureTracker(ShardId shardId) {
+        ShardIndexingPressureTracker tracker = shardIndexingPressureHotStore.get(shardId);
+        if (isNull(tracker)) {
+            // Attempt from Indexing pressure cold store
+            tracker = shardIndexingPressureColdStore.get(shardId);
+            // If not already present in cold store instantiate a new one
+            if (isNull(tracker)) {
+                ShardIndexingPressureTracker newShardIndexingPressureTracker = new ShardIndexingPressureTracker(shardId,
+                    this.shardIndexingPressureSettings.getShardPrimaryAndCoordinatingBaseLimits(),
+                    this.shardIndexingPressureSettings.getShardReplicaBaseLimits());
+                // Try update the new shard tracker to the hot store
+                tracker = shardIndexingPressureHotStore.putIfAbsent(shardId, newShardIndexingPressureTracker);
+                // Update the tracker so that we use the one actual in the hot store
+                tracker = tracker == null ? newShardIndexingPressureTracker : tracker;
+                // Write through into the cold store for future reference
+                updateShardIndexingPressureColdStore(tracker);
```

Comment:
These operations are not atomically executed
```suggestion
                tracker = shardIndexingPressureHotStore.computeIfAbsent(shardId, (k) -> {
                    ShardIndexingPressureTracker newTracker = new ShardIndexingPressureTracker(shardId, this.shardIndexingPressureSettings.getShardPrimaryAndCoordinatingBaseLimits(), this.shardIndexingPressureSettings.getShardReplicaBaseLimits());
                    updateIndexingPressureColdStore(newTracker);
                    return newTracker;
                });
```

Reply:
``computeIfAbsent`` of ``ConcurrentHashMap`` is a complex and relatively costlier operation ([synchronized](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/util/concurrent/ConcurrentHashMap.java#l1655)). It is not kept atomic, as a well thought approach here based on the use case. We need to keep the ``Get`` operations highly concurrent, and frequent operation to address sudden spike in traffic for a new shard. 
Let me know of you see any concerns with the current approach if guarantees are not being met. Will be happy to accommodate.

Reply:
https://vmlens.com/articles/scale/scalability_hash_map/ Gets are just volatile reads if that helps. See the benchmarks

Reply:
What needs a thought is that you are using a `ConcurrentHashMap` and you seem to get away without `synchronization`. You should also look into the implementation of `putIfAbsent`.

Reply:
Agree ``putIfAbsent`` on a map is unavoidable here, while the need to maintain atomicity across the two hash maps operations is avoidable.